capactio/capact
# Validate TypeInstance input

## Description

Changes proposed in this pull request:
- Validate TypeInstance input during TypeInstance creation, edition and uploading/updating TypeInstance by Argo Action.

TODO:
- [x] fix manifests - now some of TypeInstance are invalid with the Type JSON schema. (PR: https://github.com/capactio/hub-manifests/pull/56)


## Test Cases

**Test Case 1**

1. Create a sample file:
```
typeInstances:
  - alias: aws-s3
    attributes:
      - path: cap.attribute.cloud.provider.aws
        revision: 0.1.0
    typeRef:
      path: cap.type.database.object.aws.s3.bucket-config
      revision: 0.1.0
    value:
      host: https://s3.amazonaws.com
```
2. Build a binary:
```
make build-tool-cli
```
3. Verify that the validation errors occur:
```
./bin/capact-linux-amd64 typeinstances create -f test.yml 
Error: - Validation TypeInstances "TypeInstance with alias aws-s3":
    * (root): name is required
    * (root): Additional property host is not allowed
```
4. Change host to name and try one more time:
```
      name: https://s3.amazonaws.com
```
```
./bin/capact-linux-amd64 typeinstances create -f test.yml 
  ALIAS                ASSIGNED ID               
---------+---------------------------------------
  aws-s3   5605af48-c34f-4bdc-b2d8-53c679bdfa5a  
---------+---------------------------------------
```
5. The same steps could be repeated by this time with the `edit` action for TypeInstance with generated id `5605af48-c34f-4bdc-b2d8-53c679bdfa5a  `

**Test Case 2**
1. On this change try to install Matermost as described [here](https://capact.io/docs/example/mattermost-installation/).
2. Installation should not succeed on the upload TypeInstance step due to validation errors. 


## Related issue(s)
- https://github.com/capactio/capact/issues/544
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
